{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from typing import Tuple\n",
    "\n",
    "class NeedlemanWunsch:\n",
    "    \"\"\" Class for NeedlemanWunsch Alignment\n",
    "\n",
    "    Parameters:\n",
    "        sub_matrix_file: str\n",
    "            Path/filename of substitution matrix\n",
    "        gap_open: float\n",
    "            Gap opening penalty\n",
    "        gap_extend: float\n",
    "            Gap extension penalty\n",
    "\n",
    "    Attributes:\n",
    "        seqA_align: str\n",
    "            seqA alignment\n",
    "        seqB_align: str\n",
    "            seqB alignment\n",
    "        alignment_score: float\n",
    "            Score of alignment from algorithm\n",
    "        gap_open: float\n",
    "            Gap opening penalty\n",
    "        gap_extend: float\n",
    "            Gap extension penalty\n",
    "    \"\"\"\n",
    "    def __init__(self, sub_matrix_file: str, gap_open: float, gap_extend: float):\n",
    "        # Init alignment and gap matrices\n",
    "        self._align_matrix = None\n",
    "        self._gapA_matrix = None\n",
    "        self._gapB_matrix = None\n",
    "\n",
    "        # Init matrices for backtrace procedure\n",
    "        self._back = None\n",
    "        self._back_A = None\n",
    "        self._back_B = None\n",
    "\n",
    "        # Init alignment_score\n",
    "        self.alignment_score = 0\n",
    "\n",
    "        # Init empty alignment attributes\n",
    "        self.seqA_align = \"\"\n",
    "        self.seqB_align = \"\"\n",
    "\n",
    "        # Init empty sequences\n",
    "        self._seqA = \"\"\n",
    "        self._seqB = \"\"\n",
    "\n",
    "        # Setting gap open and gap extension penalties\n",
    "        self.gap_open = gap_open\n",
    "        assert gap_open < 0, \"Gap opening penalty must be negative.\"\n",
    "        self.gap_extend = gap_extend\n",
    "        assert gap_extend < 0, \"Gap extension penalty must be negative.\"\n",
    "\n",
    "        # Generating substitution matrix\n",
    "        self.sub_dict = self._read_sub_matrix(sub_matrix_file) # substitution dictionary\n",
    "\n",
    "    def _read_sub_matrix(self, sub_matrix_file):\n",
    "        \"\"\"\n",
    "        DO NOT MODIFY THIS METHOD! IT IS ALREADY COMPLETE!\n",
    "\n",
    "        This function reads in a scoring matrix from any matrix like file.\n",
    "        Where there is a line of the residues followed by substitution matrix.\n",
    "        This file also saves the alphabet list attribute.\n",
    "\n",
    "        Parameters:\n",
    "            sub_matrix_file: str\n",
    "                Name (and associated path if not in current working directory)\n",
    "                of the matrix file that contains the scoring matrix.\n",
    "\n",
    "        Returns:\n",
    "            dict_sub: dict\n",
    "                Substitution matrix dictionary with tuple of the two residues as\n",
    "                the key and score as value e.g. {('A', 'A'): 4} or {('A', 'D'): -8}\n",
    "        \"\"\"\n",
    "        with open(sub_matrix_file, 'r') as f:\n",
    "            dict_sub = {}  # Dictionary for storing scores from sub matrix\n",
    "            residue_list = []  # For storing residue list\n",
    "            start = False  # trigger for reading in score values\n",
    "            res_2 = 0  # used for generating substitution matrix\n",
    "            # reading file line by line\n",
    "            for line_num, line in enumerate(f):\n",
    "                # Reading in residue list\n",
    "                if '#' not in line.strip() and start is False:\n",
    "                    residue_list = [k for k in line.strip().upper().split(' ') if k != '']\n",
    "                    start = True\n",
    "                # Generating substitution scoring dictionary\n",
    "                elif start is True and res_2 < len(residue_list):\n",
    "                    line = [k for k in line.strip().split(' ') if k != '']\n",
    "                    # reading in line by line to create substitution dictionary\n",
    "                    assert len(residue_list) == len(line), \"Score line should be same length as residue list\"\n",
    "                    for res_1 in range(len(line)):\n",
    "                        dict_sub[(residue_list[res_1], residue_list[res_2])] = float(line[res_1])\n",
    "                    res_2 += 1\n",
    "                elif start is True and res_2 == len(residue_list):\n",
    "                    break\n",
    "        return dict_sub\n",
    "\n",
    "    def align(self, seqA: str, seqB: str) -> Tuple[float, str, str]:\n",
    "        \"\"\"\n",
    "        TODO\n",
    "\n",
    "        This function performs global sequence alignment of two strings\n",
    "        using the Needleman-Wunsch Algorithm\n",
    "\n",
    "        Parameters:\n",
    "        \tseqA: str\n",
    "         \t\tthe first string to be aligned\n",
    "         \tseqB: str\n",
    "         \t\tthe second string to be aligned with seqA\n",
    "\n",
    "        Returns:\n",
    "         \t(alignment score, seqA alignment, seqB alignment) : Tuple[float, str, str]\n",
    "         \t\tthe score and corresponding strings for the alignment of seqA and seqB\n",
    "        \"\"\"\n",
    "        # Resetting alignment in case method is called more than once\n",
    "        self.seqA_align = \"\"\n",
    "        self.seqB_align = \"\"\n",
    "\n",
    "        # Resetting alignment score in case method is called more than once\n",
    "        self.alignment_score = 0\n",
    "\n",
    "        # Initializing sequences for use in backtrace method\n",
    "        self._seqA = seqA\n",
    "        self._seqB = seqB\n",
    "\n",
    "        # TODO: Initialize matrix private attributes for use in alignment\n",
    "\n",
    "        gap_penalty = self.gap_open\n",
    "\n",
    "        sub_dict = self.sub_dict\n",
    "        # create matrices for alignment scores, gaps, and backtracing\n",
    "\n",
    "        # Determine length of each sequence and store in n or m\n",
    "        m = len(seqA)\n",
    "        n = len(seqB)\n",
    "\n",
    "        # Generate matrix of zeros to store scores\n",
    "        self._align_matrix = np.zeros((n + 1, m + 1))\n",
    "\n",
    "\n",
    "        # TODO: Implement global alignment here\n",
    "\n",
    "        for i in range(0, n + 1):\n",
    "            self._align_matrix[i][0] = gap_penalty * i\n",
    "\n",
    "        for j in range(0, m + 1):\n",
    "            self._align_matrix[0][j] = gap_penalty * j\n",
    "\n",
    "        for i in range(0, n + 1):\n",
    "            for j in range(0, m + 1):\n",
    "                match = self._align_matrix[i][j] + sub_dict[(seqA[i - 1], seqB[j - 1])]\n",
    "                delete = self._align_matrix[i][j + 1] + gap_penalty\n",
    "                insert = self._align_matrix[i + 1][j] + gap_penalty\n",
    "\n",
    "                self._align_matrix[i][j] = max(match, delete, insert)\n",
    "\n",
    "        return self._backtrace()\n",
    "\n",
    "    def _backtrace(self) -> Tuple[float, str, str]:\n",
    "        \"\"\"\n",
    "        TODO\n",
    "\n",
    "        This function traces back through the back matrix created with the\n",
    "        align function in order to return the final alignment score and strings.\n",
    "\n",
    "        Parameters:\n",
    "        \tNone\n",
    "\n",
    "        Returns:\n",
    "         \t(alignment score, seqA alignment, seqB alignment) : Tuple[float, str, str]\n",
    "         \t\tthe score and corresponding strings for the alignment of seqA and seqB\n",
    "        \"\"\"\n",
    "        seqA = self._seqA\n",
    "        seqB = self._seqB\n",
    "        sub_dict = self.sub_dict\n",
    "        alignA = self.seqA_align\n",
    "        alignB = self.seqB_align\n",
    "\n",
    "        gap_penalty = self.gap_open\n",
    "\n",
    "        i = len(seqA)\n",
    "        j = len(seqB)\n",
    "\n",
    "        while i > 0 and j > 0:  # end touching the top or the left edge\n",
    "            score_current = self._align_matrix[i][j]\n",
    "            score_diagonal = self._align_matrix[i - 1][j - 1]\n",
    "            score_up = self._align_matrix[i][j - 1]\n",
    "            score_left = self._align_matrix[i - 1][j]\n",
    "\n",
    "            # Check to figure out which cell the current score was calculated from,\n",
    "            # then update i and j to correspond to that cell.\n",
    "            if score_current == score_diagonal + sub_dict[(seqA[j - 1], seqB[i - 1])]:\n",
    "                alignA += seqA[j - 1]\n",
    "                alignB += seqB[i - 1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif score_current == score_up + gap_penalty:\n",
    "                alignA += seqA[j - 1]\n",
    "                alignB += '-'\n",
    "                j -= 1\n",
    "            elif score_current == score_left + gap_penalty:\n",
    "                alignA += '-'\n",
    "                alignB += seqB[i - 1]\n",
    "                i -= 1\n",
    "\n",
    "        # Finish tracing up to the top left cell\n",
    "        while i > 0:\n",
    "            alignA += seqA[i - 1]\n",
    "            alignB += '-'\n",
    "            j -= 1\n",
    "        while j > 0:\n",
    "            alignA += '-'\n",
    "            alignB += seqB[j - 1]\n",
    "            i -= 1\n",
    "\n",
    "        # Traversed the score matrix from the bottom right so the two sequences will be reversed.\n",
    "        # Reverse the sequences.\n",
    "        alignA = alignA[::-1]\n",
    "        alignB = alignB[::-1]\n",
    "\n",
    "        # Update seq alignment values\n",
    "        self.seqA_align = alignA\n",
    "        self.seqB_align = alignB\n",
    "\n",
    "        self.alignment_score = self._align_matrix.sum()\n",
    "\n",
    "        return (self.alignment_score, self.seqA_align, self.seqB_align)\n",
    "\n",
    "\n",
    "def read_fasta(fasta_file: str) -> Tuple[str, str]:\n",
    "    \"\"\"\n",
    "    DO NOT MODIFY THIS FUNCTION! IT IS ALREADY COMPLETE!\n",
    "\n",
    "    This function reads in a FASTA file and returns the associated\n",
    "    string of characters (residues or nucleotides) and the header.\n",
    "    This function assumes a single protein or nucleotide sequence\n",
    "    per fasta file and will only read in the first sequence in the\n",
    "    file if multiple are provided.\n",
    "\n",
    "    Parameters:\n",
    "        fasta_file: str\n",
    "            name (and associated path if not in current working directory)\n",
    "            of the Fasta file.\n",
    "\n",
    "    Returns:\n",
    "        seq: str\n",
    "            String of characters from FASTA file\n",
    "        header: str\n",
    "            Fasta header\n",
    "    \"\"\"\n",
    "    assert fasta_file.endswith(\".fa\"), \"Fasta file must be a fasta file with the suffix .fa\"\n",
    "    with open(fasta_file) as f:\n",
    "        seq = \"\"  # initializing sequence\n",
    "        first_header = True\n",
    "        for line in f:\n",
    "            is_header = line.strip().startswith(\">\")\n",
    "            # Reading in the first header\n",
    "            if is_header and first_header:\n",
    "                header = line.strip()  # reading in fasta header\n",
    "                first_header = False\n",
    "            # Reading in the sequence line by line\n",
    "            elif not is_header:\n",
    "                seq += line.strip().upper()  # generating full sequence\n",
    "            # Breaking if more than one header is provided in the fasta file\n",
    "            elif is_header and not first_header:\n",
    "                break\n",
    "    return seq, header\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "alg = NeedlemanWunsch(\"substitution_matrices/BLOSUM62.mat\", -10.0, -1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1, _ = read_fasta(\"./data/test_seq3.fa\")\n",
    "seq2, _ = read_fasta(\"./data/test_seq4.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[  0. -10. -20. -30. -40. -50. -60.]\n",
      " [-10.  14.   1.  -8. -13. -26. -31.]\n",
      " [-20.   1.  19.   8.   2.   2.   4.]\n",
      " [-30.  -8.   1.  17.   8.  -1.   2.]\n",
      " [-40. -17.   1.   7.  17.  17.  -1.]\n",
      " [-50. -26.   8.   6.   7.  16.  17.]]\n"
     ]
    }
   ],
   "source": [
    "n = len(seq3)\n",
    "m = len(seq4)\n",
    "gap_open_penalty = -10\n",
    "gap_ext_penalty = -1\n",
    "\n",
    "# Initialize empty matrices\n",
    "score_mat = np.zeros((m + 1, n + 1))\n",
    "align_mat = np.zeros((m + 1, n + 1))\n",
    "gapA_mat = np.zeros((m + 1, n + 1))\n",
    "gapB_mat = np.zeros((m + 1, n + 1))\n",
    "\n",
    "# Initialize matrix values\n",
    "\n",
    "for i in range(0, m + 1):\n",
    "    score_mat[i][0] = gap_open_penalty * i\n",
    "    align_mat[i][0] = gap_open_penalty * i\n",
    "    gapA_mat[i][0] = np.NINF\n",
    "    gapB_mat[i][0] = -(gap_open_penalty + gap_ext_penalty * i)\n",
    "\n",
    "for j in range(0, n + 1):\n",
    "    score_mat[0][j] = gap_open_penalty * j\n",
    "    align_mat[0][j] = gap_open_penalty * j\n",
    "    gapA_mat[0][j] = -(gap_open_penalty + gap_ext_penalty * j)\n",
    "    gapB_mat[0][j] = np.NINF\n",
    "    \n",
    "for i in range(1, m + 1):\n",
    "    for j in range(1, n + 1):\n",
    "        score_mat[i][j] = (align_mat[i - 1][j - 1] + alg.sub_dict[(seq3[j-1], seq4[i-1])]) + max(score_mat[i-1][j-1], gapA_mat[i-1][j-1], gapB_mat[i-1][j-1])\n",
    "        gapA_mat[i][j] = max((score_mat[i][j - 1] + (gap_open_penalty + gap_ext_penalty)), (gapA_mat[i][j-1] + gap_open_penalty), ((gapB_mat[i][j-1]) + (gap_open_penalty+gap_ext_penalty)))\n",
    "        gapB_mat[i][j] = max((score_mat[i - 1][j] + (gap_open_penalty + gap_ext_penalty)), (gapA_mat[i - 1][j] + (gap_open_penalty + gap_ext_penalty)), (gapB_mat[i - 1][j] + gap_open_penalty))\n",
    "print(score_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 10.  11.  12.  13.  14.  15.  16.]\n",
      " [-inf   0.   3.  -7.  -9.  -8.  -7.]\n",
      " [-inf   1.  -8.   8.  -2.  -9.  -9.]\n",
      " [-inf   2.  -8.  -3.   6.  -3. -12.]\n",
      " [-inf   3.  -7. -10.  -4.   6.   6.]\n",
      " [-inf   4.  -6.  -3.  -5.  -4.   5.]]\n"
     ]
    }
   ],
   "source": [
    "print(gapA_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[-inf -inf -inf -inf -inf -inf -inf]\n",
      " [ 11.   0.   1.   2.   3.   4.   5.]\n",
      " [ 12.   3.  -8.  -8.  -7.  -6.  -5.]\n",
      " [ 13.  -7.   8.  -3.  -9.  -9.  -7.]\n",
      " [ 14.  -9.  -2.   6.  -3. -12.  -9.]\n",
      " [ 15.  -8. -10.  -4.   6.   6.  -5.]]\n"
     ]
    }
   ],
   "source": [
    "print(gapB_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "string index out of range",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mIndexError\u001B[0m                                Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[20], line 21\u001B[0m\n\u001B[1;32m     18\u001B[0m \u001B[38;5;66;03m# Check to figure out which cell the current score was calculated from,\u001B[39;00m\n\u001B[1;32m     19\u001B[0m \u001B[38;5;66;03m# then update i and j to correspond to that cell.\u001B[39;00m\n\u001B[1;32m     20\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m score_current \u001B[38;5;241m==\u001B[39m match:\n\u001B[0;32m---> 21\u001B[0m     alignA \u001B[38;5;241m+\u001B[39m\u001B[38;5;241m=\u001B[39m \u001B[43mseq1\u001B[49m\u001B[43m[\u001B[49m\u001B[43mj\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m-\u001B[39;49m\u001B[43m \u001B[49m\u001B[38;5;241;43m1\u001B[39;49m\u001B[43m]\u001B[49m\n\u001B[1;32m     22\u001B[0m     alignB \u001B[38;5;241m+\u001B[39m\u001B[38;5;241m=\u001B[39m seq2[i \u001B[38;5;241m-\u001B[39m \u001B[38;5;241m1\u001B[39m]\n\u001B[1;32m     23\u001B[0m     i \u001B[38;5;241m-\u001B[39m\u001B[38;5;241m=\u001B[39m \u001B[38;5;241m1\u001B[39m\n",
      "\u001B[0;31mIndexError\u001B[0m: string index out of range"
     ]
    }
   ],
   "source": [
    "i = m\n",
    "j = n\n",
    "\n",
    "alignA = \"\"\n",
    "alignB = \"\"\n",
    "\n",
    "while i > 0 and j > 0:  # end touching the top or the left edge\n",
    "    score_current = score_mat[i][j]\n",
    "    score_diagonal = score_mat[i - 1][j - 1]\n",
    "    score_up = score_mat[i][j - 1]\n",
    "    score_left = score_mat[i - 1][j]\n",
    "\n",
    "    # Check to figure out which cell the current score was calculated from,\n",
    "    # then update i and j to correspond to that cell.\n",
    "    if score_current == match:\n",
    "        alignA += seq1[j - 1]\n",
    "        alignB += seq2[i - 1]\n",
    "        i -= 1\n",
    "        j -= 1\n",
    "    elif score_current == gapA_mat[i, j]:\n",
    "        alignA += seq1[j - 1]\n",
    "        alignB += '-'\n",
    "        j -= 1\n",
    "    elif score_current == gapB_mat[i, j]:\n",
    "        alignA += '-'\n",
    "        alignB += seq2[i - 1]\n",
    "        i -= 1\n",
    "\n",
    "# Finish tracing up to the top left cell\n",
    "while i > 0:\n",
    "    alignA += seq1[i - 1]\n",
    "    alignB += '-'\n",
    "    j -= 1\n",
    "while j > 0:\n",
    "    alignA += '-'\n",
    "    alignB += seq2[j - 1]\n",
    "    i -= 1\n",
    "\n",
    "# Traversed the score matrix from the bottom right so the two sequences will be reversed.\n",
    "# Reverse the sequences.\n",
    "alignA = alignA[::-1]\n",
    "alignB = alignB[::-1]\n",
    "\n",
    "# Update seq alignment values\n",
    "print(alignA)\n",
    "print(alignB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "score_current"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
