{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b0783f5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from typing import Tuple\n",
    "\n",
    "class NeedlemanWunsch:\n",
    "    \"\"\" Class for NeedlemanWunsch Alignment\n",
    "\n",
    "    Parameters:\n",
    "        sub_matrix_file: str\n",
    "            Path/filename of substitution matrix\n",
    "        gap_open: float\n",
    "            Gap opening penalty\n",
    "        gap_extend: float\n",
    "            Gap extension penalty\n",
    "\n",
    "    Attributes:\n",
    "        seqA_align: str\n",
    "            seqA alignment\n",
    "        seqB_align: str\n",
    "            seqB alignment\n",
    "        alignment_score: float\n",
    "            Score of alignment from algorithm\n",
    "        gap_open: float\n",
    "            Gap opening penalty\n",
    "        gap_extend: float\n",
    "            Gap extension penalty\n",
    "    \"\"\"\n",
    "    def __init__(self, sub_matrix_file: str, gap_open: float, gap_extend: float):\n",
    "        # Init alignment and gap matrices\n",
    "        self._align_matrix = None\n",
    "        self._gapA_matrix = None\n",
    "        self._gapB_matrix = None\n",
    "\n",
    "        # Init matrices for backtrace procedure\n",
    "        self._back = None\n",
    "        self._back_A = None\n",
    "        self._back_B = None\n",
    "\n",
    "        # Init alignment_score\n",
    "        self.alignment_score = 0\n",
    "\n",
    "        # Init empty alignment attributes\n",
    "        self.seqA_align = \"\"\n",
    "        self.seqB_align = \"\"\n",
    "\n",
    "        # Init empty sequences\n",
    "        self._seqA = \"\"\n",
    "        self._seqB = \"\"\n",
    "\n",
    "        # Setting gap open and gap extension penalties\n",
    "        self.gap_open = gap_open\n",
    "        assert gap_open < 0, \"Gap opening penalty must be negative.\"\n",
    "        self.gap_extend = gap_extend\n",
    "        assert gap_extend < 0, \"Gap extension penalty must be negative.\"\n",
    "\n",
    "        # Generating substitution matrix\n",
    "        self.sub_dict = self._read_sub_matrix(sub_matrix_file) # substitution dictionary\n",
    "\n",
    "    def _read_sub_matrix(self, sub_matrix_file):\n",
    "        \"\"\"\n",
    "        DO NOT MODIFY THIS METHOD! IT IS ALREADY COMPLETE!\n",
    "\n",
    "        This function reads in a scoring matrix from any matrix like file.\n",
    "        Where there is a line of the residues followed by substitution matrix.\n",
    "        This file also saves the alphabet list attribute.\n",
    "\n",
    "        Parameters:\n",
    "            sub_matrix_file: str\n",
    "                Name (and associated path if not in current working directory)\n",
    "                of the matrix file that contains the scoring matrix.\n",
    "\n",
    "        Returns:\n",
    "            dict_sub: dict\n",
    "                Substitution matrix dictionary with tuple of the two residues as\n",
    "                the key and score as value e.g. {('A', 'A'): 4} or {('A', 'D'): -8}\n",
    "        \"\"\"\n",
    "        with open(sub_matrix_file, 'r') as f:\n",
    "            dict_sub = {}  # Dictionary for storing scores from sub matrix\n",
    "            residue_list = []  # For storing residue list\n",
    "            start = False  # trigger for reading in score values\n",
    "            res_2 = 0  # used for generating substitution matrix\n",
    "            # reading file line by line\n",
    "            for line_num, line in enumerate(f):\n",
    "                # Reading in residue list\n",
    "                if '#' not in line.strip() and start is False:\n",
    "                    residue_list = [k for k in line.strip().upper().split(' ') if k != '']\n",
    "                    start = True\n",
    "                # Generating substitution scoring dictionary\n",
    "                elif start is True and res_2 < len(residue_list):\n",
    "                    line = [k for k in line.strip().split(' ') if k != '']\n",
    "                    # reading in line by line to create substitution dictionary\n",
    "                    assert len(residue_list) == len(line), \"Score line should be same length as residue list\"\n",
    "                    for res_1 in range(len(line)):\n",
    "                        dict_sub[(residue_list[res_1], residue_list[res_2])] = float(line[res_1])\n",
    "                    res_2 += 1\n",
    "                elif start is True and res_2 == len(residue_list):\n",
    "                    break\n",
    "        return dict_sub\n",
    "\n",
    "    def align(self, seqA: str, seqB: str) -> Tuple[float, str, str]:\n",
    "        \"\"\"\n",
    "        TODO\n",
    "\n",
    "        This function performs global sequence alignment of two strings\n",
    "        using the Needleman-Wunsch Algorithm\n",
    "\n",
    "        Parameters:\n",
    "        \tseqA: str\n",
    "         \t\tthe first string to be aligned\n",
    "         \tseqB: str\n",
    "         \t\tthe second string to be aligned with seqA\n",
    "\n",
    "        Returns:\n",
    "         \t(alignment score, seqA alignment, seqB alignment) : Tuple[float, str, str]\n",
    "         \t\tthe score and corresponding strings for the alignment of seqA and seqB\n",
    "        \"\"\"\n",
    "        # Resetting alignment in case method is called more than once\n",
    "        self.seqA_align = \"\"\n",
    "        self.seqB_align = \"\"\n",
    "\n",
    "        # Resetting alignment score in case method is called more than once\n",
    "        self.alignment_score = 0\n",
    "\n",
    "        # Initializing sequences for use in backtrace method\n",
    "        self._seqA = seqA\n",
    "        self._seqB = seqB\n",
    "\n",
    "        # TODO: Initialize matrix private attributes for use in alignment\n",
    "\n",
    "        gap_penalty = self.gap_open\n",
    "\n",
    "        sub_dict = self.sub_dict\n",
    "        # create matrices for alignment scores, gaps, and backtracing\n",
    "\n",
    "        # Determine length of each sequence and store in n or m\n",
    "        m = len(seqA)\n",
    "        n = len(seqB)\n",
    "\n",
    "        # Generate matrix of zeros to store scores\n",
    "        self._align_matrix = np.zeros((n + 1, m + 1))\n",
    "\n",
    "\n",
    "        # TODO: Implement global alignment here\n",
    "\n",
    "        for i in range(0, n + 1):\n",
    "            self._align_matrix[i][0] = gap_penalty * i\n",
    "\n",
    "        for j in range(0, m + 1):\n",
    "            self._align_matrix[0][j] = gap_penalty * j\n",
    "\n",
    "        for i in range(0, n + 1):\n",
    "            for j in range(0, m + 1):\n",
    "                match = self._align_matrix[i][j] + sub_dict[(seqA[i - 1], seqB[j - 1])]\n",
    "                delete = self._align_matrix[i][j + 1] + gap_penalty\n",
    "                insert = self._align_matrix[i + 1][j] + gap_penalty\n",
    "\n",
    "                self._align_matrix[i][j] = max(match, delete, insert)\n",
    "\n",
    "        return self._backtrace()\n",
    "\n",
    "    def _backtrace(self) -> Tuple[float, str, str]:\n",
    "        \"\"\"\n",
    "        TODO\n",
    "\n",
    "        This function traces back through the back matrix created with the\n",
    "        align function in order to return the final alignment score and strings.\n",
    "\n",
    "        Parameters:\n",
    "        \tNone\n",
    "\n",
    "        Returns:\n",
    "         \t(alignment score, seqA alignment, seqB alignment) : Tuple[float, str, str]\n",
    "         \t\tthe score and corresponding strings for the alignment of seqA and seqB\n",
    "        \"\"\"\n",
    "        seqA = self._seqA\n",
    "        seqB = self._seqB\n",
    "        sub_dict = self.sub_dict\n",
    "        alignA = self.seqA_align\n",
    "        alignB = self.seqB_align\n",
    "\n",
    "        gap_penalty = self.gap_open\n",
    "\n",
    "        i = len(seqA)\n",
    "        j = len(seqB)\n",
    "\n",
    "        while i > 0 and j > 0:  # end touching the top or the left edge\n",
    "            score_current = self._align_matrix[i][j]\n",
    "            score_diagonal = self._align_matrix[i - 1][j - 1]\n",
    "            score_up = self._align_matrix[i][j - 1]\n",
    "            score_left = self._align_matrix[i - 1][j]\n",
    "\n",
    "            # Check to figure out which cell the current score was calculated from,\n",
    "            # then update i and j to correspond to that cell.\n",
    "            if score_current == score_diagonal + sub_dict[(seqA[j - 1], seqB[i - 1])]:\n",
    "                alignA += seqA[j - 1]\n",
    "                alignB += seqB[i - 1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif score_current == score_up + gap_penalty:\n",
    "                alignA += seqA[j - 1]\n",
    "                alignB += '-'\n",
    "                j -= 1\n",
    "            elif score_current == score_left + gap_penalty:\n",
    "                alignA += '-'\n",
    "                alignB += seqB[i - 1]\n",
    "                i -= 1\n",
    "\n",
    "        # Finish tracing up to the top left cell\n",
    "        while i > 0:\n",
    "            alignA += seqA[i - 1]\n",
    "            alignB += '-'\n",
    "            j -= 1\n",
    "        while j > 0:\n",
    "            alignA += '-'\n",
    "            alignB += seqB[j - 1]\n",
    "            i -= 1\n",
    "\n",
    "        # Traversed the score matrix from the bottom right so the two sequences will be reversed.\n",
    "        # Reverse the sequences.\n",
    "        alignA = alignA[::-1]\n",
    "        alignB = alignB[::-1]\n",
    "\n",
    "        # Update seq alignment values\n",
    "        self.seqA_align = alignA\n",
    "        self.seqB_align = alignB\n",
    "\n",
    "        self.alignment_score = self._align_matrix.sum()\n",
    "\n",
    "        return (self.alignment_score, self.seqA_align, self.seqB_align)\n",
    "\n",
    "\n",
    "def read_fasta(fasta_file: str) -> Tuple[str, str]:\n",
    "    \"\"\"\n",
    "    DO NOT MODIFY THIS FUNCTION! IT IS ALREADY COMPLETE!\n",
    "\n",
    "    This function reads in a FASTA file and returns the associated\n",
    "    string of characters (residues or nucleotides) and the header.\n",
    "    This function assumes a single protein or nucleotide sequence\n",
    "    per fasta file and will only read in the first sequence in the\n",
    "    file if multiple are provided.\n",
    "\n",
    "    Parameters:\n",
    "        fasta_file: str\n",
    "            name (and associated path if not in current working directory)\n",
    "            of the Fasta file.\n",
    "\n",
    "    Returns:\n",
    "        seq: str\n",
    "            String of characters from FASTA file\n",
    "        header: str\n",
    "            Fasta header\n",
    "    \"\"\"\n",
    "    assert fasta_file.endswith(\".fa\"), \"Fasta file must be a fasta file with the suffix .fa\"\n",
    "    with open(fasta_file) as f:\n",
    "        seq = \"\"  # initializing sequence\n",
    "        first_header = True\n",
    "        for line in f:\n",
    "            is_header = line.strip().startswith(\">\")\n",
    "            # Reading in the first header\n",
    "            if is_header and first_header:\n",
    "                header = line.strip()  # reading in fasta header\n",
    "                first_header = False\n",
    "            # Reading in the sequence line by line\n",
    "            elif not is_header:\n",
    "                seq += line.strip().upper()  # generating full sequence\n",
    "            # Breaking if more than one header is provided in the fasta file\n",
    "            elif is_header and not first_header:\n",
    "                break\n",
    "    return seq, header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "115cbb5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "alg = NeedlemanWunsch(\"substitution_matrices/BLOSUM62.mat\", -10.0, -1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "b0b93d91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17.0\n"
     ]
    }
   ],
   "source": [
    "seq1, _ = read_fasta(\"./data/test_seq3.fa\")\n",
    "seq2, _ = read_fasta(\"./data/test_seq4.fa\")\n",
    "n = len(seq1) # 10\n",
    "m = len(seq2) # 7\n",
    "gap_open_penalty = -10\n",
    "gap_ext_penalty = -1\n",
    "\n",
    "# Initialize empty matrices\n",
    "align_mat = np.zeros((m + 1, n + 1)) # rows vs col\n",
    "score_mat = np.zeros((m + 1, n + 1))\n",
    "gapA_mat = np.zeros((m + 1, n + 1))\n",
    "gapB_mat = np.zeros((m + 1, n + 1))\n",
    "\n",
    "for i in range(1, m + 1):\n",
    "    align_mat[i][0] = -np.inf\n",
    "    gapA_mat[i][0] = gap_open_penalty + ((gap_ext_penalty) * i)\n",
    "    gapB_mat[i][0] = -np.inf\n",
    "    \n",
    "for j in range(1, n + 1):\n",
    "    align_mat[0][j] = -np.inf\n",
    "    gapA_mat[0][j] = -np.inf\n",
    "    gapB_mat[0][j] = gap_open_penalty + ((gap_ext_penalty) * j)\n",
    "\n",
    "for i in range(1, m + 1):\n",
    "    for j in range(1, n + 1):\n",
    "        align_mat[i][j] = max(align_mat[i - 1][j - 1] + alg.sub_dict[(seq1[j-1], seq2[i-1])], gapA_mat[i-1][j-1] + alg.sub_dict[(seq1[j-1], seq2[i-1])], gapB_mat[i-1][j-1] + alg.sub_dict[(seq1[j-1], seq2[i-1])])\n",
    "        gapA_mat[i][j] = max((align_mat[i - 1][j] + gap_open_penalty + gap_ext_penalty), (gapA_mat[i-1][j] + (gap_ext_penalty)))\n",
    "        gapB_mat[i][j] = max((align_mat[i][j - 1] + gap_open_penalty + gap_ext_penalty), (gapB_mat[i][j-1] + (gap_ext_penalty)))\n",
    "        score_mat[i][j] = max(align_mat[i][j], gapA_mat[i][j], gapB_mat[i][j])\n",
    "print(np.max(score_mat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "f41c6485",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[  0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.]\n",
      " [  0.   5.  -6.  -7.  -8.  -9. -10. -11. -12. -13. -14.]\n",
      " [  0.  -6.   4.  -7.  -7.  -3. -10. -11. -10. -11. -14.]\n",
      " [  0.  -7.  -7.   5.  -6.  -7.   1.  -8. -10. -11. -12.]\n",
      " [  0.  -8.  -8.  -4.   2.  -9.  -5.   5.  -6.  -7.  -8.]\n",
      " [  0.  -9.  -9.  -7.  -4.   3.  -8.  -6.  10.  -1.  -2.]\n",
      " [  0. -10. -10.  -8.   1.  -4.   0.  -7.  -1.  10.  -1.]\n",
      " [  0. -11. -11.  -9. -10.   0.  -7.  -3.  -2.  -1.  17.]]\n"
     ]
    }
   ],
   "source": [
    "print(score_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "b1dda9da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[  0. -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf]\n",
      " [-inf   5. -12. -11. -15. -14. -13. -15. -18. -19. -21.]\n",
      " [-inf -11.   4.  -8.  -7.  -3. -11. -13. -10. -11. -14.]\n",
      " [-inf -10.  -7.   5. -10.  -9.   1.  -8. -13. -12. -14.]\n",
      " [-inf -12.  -8.  -4.   2.  -9.  -5.   5. -11. -13. -14.]\n",
      " [-inf -15.  -9. -11.  -4.   3. -11.  -8.  10.  -1.  -9.]\n",
      " [-inf -17. -11. -12.   1.  -4.   0. -11.  -6.  10.  -3.]\n",
      " [-inf -18. -11. -12. -10.   0.  -7.  -3.  -9.  -3.  17.]]\n"
     ]
    }
   ],
   "source": [
    "print(align_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "1a606f9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill in back matrix for backtracing\n",
    "traceback_mat = np.ones((m + 1, n + 1)) * -np.inf\n",
    "#best_score = max(align_mat[i][j], gapA_mat[i][j], gapB_mat[i][j])\n",
    "\n",
    "for i in range(1, m + 1):\n",
    "    for j in range(1, n + 1):\n",
    "        # A match is represented in the back matrix as a 0\n",
    "        if score_mat[i][j] == align_mat[i][j]:\n",
    "            traceback_mat[i][j] = 0\n",
    "        # A gap in seqA is represented in the back matrix as a -1\n",
    "        elif score_mat[i][j] == gapA_mat[i][j]:\n",
    "            traceback_mat[i][j] = -1\n",
    "        # A gap in seqB is represented in the back matrix as a 1\n",
    "        else:\n",
    "            traceback_mat[i][j] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "50644523",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[-inf -inf -inf -inf -inf -inf -inf -inf -inf -inf -inf]\n",
      " [-inf   0.   1.   1.   1.   1.   1.   1.   1.   1.   1.]\n",
      " [-inf  -1.   0.   1.   0.   0.   1.   1.   0.   0.   0.]\n",
      " [-inf  -1.   0.   0.   1.   1.   0.   0.   1.   1.   1.]\n",
      " [-inf  -1.   0.   0.   0.   0.   0.   0.   1.   1.   1.]\n",
      " [-inf  -1.   0.  -1.   0.   0.   1.  -1.   0.   0.   1.]\n",
      " [-inf  -1.  -1.  -1.   0.   0.   0.  -1.  -1.   0.   1.]\n",
      " [-inf  -1.   0.  -1.   0.   0.   0.   0.  -1.  -1.   0.]]\n"
     ]
    }
   ],
   "source": [
    "print(traceback_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "5ff6da54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAVHQLIRRP\n",
      "MQLIRHP\n"
     ]
    }
   ],
   "source": [
    "print(seq1)\n",
    "print(seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "e0709a0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqA_align = \"\"\n",
    "seqB_align = \"\"\n",
    "\n",
    "while i > 0 and j > 0:\n",
    "    back_step = traceback_mat[i][j]\n",
    "    if back_step == 0:\n",
    "        seqA_align += seq1[j-1]\n",
    "        seqB_align += seq2[i-1]\n",
    "        i -= 1\n",
    "        j -= 1\n",
    "    elif back_step == -1:\n",
    "        seqA_align += \"-\"\n",
    "        seqB_align = seq2[i -1]\n",
    "        i-=1\n",
    "    elif back_step == 1:\n",
    "        seqA_align += seq1[j - 1]\n",
    "        seqB_align += \"-\"\n",
    "        j -= 1\n",
    "\n",
    "if i > 0 and j == 0:\n",
    "    while i > 0:\n",
    "        seqA_align += '-'\n",
    "        seqB_align += seq2[j-1]\n",
    "        i -= 1\n",
    "if i == 0 and j > 0:\n",
    "    while j > 0:\n",
    "        seqA_align += seq1[j - 1]\n",
    "        seqB_align += '-'\n",
    "        i -= 1\n",
    "    \n",
    "seqA_align = seqA_align[::-1]\n",
    "seqB_align = seqB_align[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "62f3e5ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAVHQLIRRP\n",
      "M---QLIRHP\n"
     ]
    }
   ],
   "source": [
    "print(seqA_align)\n",
    "print(seqB_align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "a14a6894",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "356cebf2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
